controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -320,51 +323,66 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
         CompletableFuture<Void> validate = validate(lease, scaleGracePeriod);
         long maxExecutionPeriod = Math.min(MAX_EXECUTION_TIME_MULTIPLIER * lease, Duration.ofDays(1).toMillis());
 
-        UUID txnId = UUID.randomUUID();
-        TxnResource resource = new TxnResource(scope, stream, txnId);
-
-        // Step 2. Add txn to host-transaction index.
-        CompletableFuture<Void> addIndex = validate.thenComposeAsync(ignore ->
-                streamMetadataStore.addTxnToIndex(hostId, resource, 0), executor).whenComplete((v, e) -> {
-                    if (e != null) {
-                        log.debug("Txn={}, failed adding txn to host-txn index of host={}", txnId, hostId);
-                    } else {
-                        log.debug("Txn={}, added txn to host-txn index of host={}", txnId, hostId);
-                    }
-                });
-
-        // Step 3. Create txn node in the store.
-        CompletableFuture<VersionedTransactionData> txnFuture = addIndex.thenComposeAsync(ignore ->
-                streamMetadataStore.createTransaction(scope, stream, txnId, lease, maxExecutionPeriod,
-                        scaleGracePeriod, ctx, executor), executor).whenComplete((v, e) -> {
-                    if (e != null) {
-                        log.debug("Txn={}, failed creating txn in store", txnId);
-                    } else {
-                        log.debug("Txn={}, created in store", txnId);
-                    }
-                });
+        // 1. get latest epoch from history
+        // 2. generateNewTransactionId.. this step can throw WriteConflictException
+        // 3. txn id = 32 bit epoch + 96 bit counter
+        // 4. if while creating txn epoch no longer exists, then we will get DataNotFoundException.
```

Comment:
I think step 4 is saying that in the case the epoch changes between steps 1 and 4 say due to a concurrent execution, the txn creation will fail and will force the controller to start over with the txn creation, including generating a new id. Is that it?

Reply:
yes thats correct. 
however, this is only temporary.. with rolling transaction implementation, we will never run into such situations as old epochs will be allowed to exist.